ground/src/main/java/com/google/android/ground/persistence/sync/LocalMutationSyncWorker.kt

```diff
@@ -82,63 +83,55 @@ constructor(
-      val user = getUser(userId)
-      if (mutations == null || user == null) {
-        continue
-      }
-      val result = processMutations(mutations, user)
-      if (!result) {
-        noErrors = false
-      }
-    }
-    return noErrors
-  }
-
   /**
    * Applies mutations to remote data store. Once successful, removes them from the local db.
    *
-   * @return `true` if the mutations were successfully synced with [RemoteDataStore].
+   * @return `success` if the mutations were successfully synced with [RemoteDataStore].
    */
-  private suspend fun processMutations(mutations: List<Mutation>, user: User): Boolean {
-    check(mutations.isNotEmpty()) { "List of mutations is empty" }
-
+  private suspend fun processMutations(mutations: List<Mutation>): Result {
+    if (mutations.isEmpty()) return success()
+    val user = getUserFromMutations(mutations) ?: return failure()
     return try {
       mutationRepository.markAsInProgress(mutations)
       remoteDataStore.applyMutations(mutations, user)
       mutationRepository.finalizePendingMutationsForMediaUpload(mutations)
-      true
+      success()
     } catch (t: Throwable) {
-      // Mark all mutations as having failed since the remote datastore only commits when all
-      // mutations have succeeded.
       Timber.d(t, "Local mutation sync failed")
-      val crashlytics = FirebaseCrashLogger()
-      crashlytics.setSelectedSurveyId(mutations.first().surveyId)
-      crashlytics.logException(t)
-      mutationRepository.markAsFailed(mutations, t)
-      false
+      handleMutationSyncFailed(mutations, t)
+      retry()
     }
   }
 
-  private suspend fun getUser(userId: String): User? {
-    val user = localUserStore.getUserOrNull(userId)
+  /**
+   * Returns the user associated with the mutations. If missing, removes the mutations from local
+   * storage as well.
+   */
+  private suspend fun getUserFromMutations(mutations: List<Mutation>): User? {
+    val userIds = mutations.map { it.userId }.toSet()
+
+    check(userIds.size == 1) { "Found mutations for multiple users" }
+    // TODO: Add a check to assert that the user id matches the currently logged in user.
+
+    val user = localUserStore.getUserOrNull(userIds.first())
     if (user == null) {
-      Timber.e("User account removed before mutation processed")
+      Timber.e("User removed before mutation processed. Removing %d mutations", mutations.size)
+      mutationRepository.finalizeDeletions(mutations)
     }
     return user
   }
 
+  /**
+   * Marks all mutations as failed since the remote datastore only commits when all mutations have
+   * succeeded.
+   */
+  private suspend fun handleMutationSyncFailed(mutations: List<Mutation>, throwable: Throwable) {
+    mutationRepository.markAsFailed(mutations, throwable)
+    val crashlytics = FirebaseCrashLogger()
+    crashlytics.setSelectedSurveyId(mutations.first().surveyId)
+    crashlytics.logException(throwable)
```

Comment:
I noticed this earlier - imo we should update `logException` to require a custom message to help us identify the source of these errors more easily in Crashlytics.

Reply:
Why aren't we using `Timber.e` instead? That would tag the exception and log to crashlytics.

Reply:
Does it? How does that work?

Reply:
Timber.e() reaches this method.

https://github.com/google/ground-android/blob/17422bdd7f9b3cd1b8bdec94f0a6f1f4a10a5245/ground/src/main/java/com/google/android/ground/GroundApplication.kt#L64

which then logs the exception to crashlytics if the log level is Error

https://github.com/google/ground-android/blob/17422bdd7f9b3cd1b8bdec94f0a6f1f4a10a5245/ground/src/main/java/com/google/android/ground/FirebaseCrashLogger.kt#L32